File: mmoTest/Assets/Script/Data/Proto/Role/Send/RoleOperation_CreateRoleProto.cs
//===================================================
//作    者：边涯  http://www.u3dol.com  QQ群：87481002
//创建时间：2020-03-15 10:58:27
//备    注：
//===================================================
using System.Collections;
using System.Collections.Generic;
using System;

/// <summary>
/// 客户端发送创建角色消息
/// </summary>
public struct RoleOperation_CreateRoleProto : IProto
{
    public ushort ProtoCode { get { return 10003; } }

    public byte JobId; //职业ID
    public string RoleNickName; //角色名称

    public byte[] ToArray()
    {
        using (MMO_MemoryStream ms = new MMO_MemoryStream())
        {
            ms.WriteUShort(ProtoCode);
            ms.WriteByte(JobId);
            ms.WriteUTF8String(RoleNickName);
            return ms.ToArray();
        }
    }

    public static RoleOperation_CreateRoleProto GetProto(byte[] buffer)
    {
        RoleOperation_CreateRoleProto proto = new RoleOperation_CreateRoleProto();
        using (MMO_MemoryStream ms = new MMO_MemoryStream(buffer))
        {
            proto.JobId = (byte)ms.ReadByte();
            proto.RoleNickName = ms.ReadUTF8String();
        }
        return proto;
    }
}

File: MMORPGServer/Proto/CreateLuaProto/RoleOperation_CreateRoleProto.lua
--客户端发送创建角色消息
RoleOperation_CreateRoleProto = { ProtoCode = 10003, JobId = 0, RoleNickName = "" }

--这句是重定义元表的索引，就是说有了这句，这个才是一个类
RoleOperation_CreateRoleProto.__index = RoleOperation_CreateRoleProto;

function RoleOperation_CreateRoleProto.New()
    local self = { }; --初始化self
    setmetatable(self, RoleOperation_CreateRoleProto); --将self的元表设定为Class
    return self;
end


--发送协议
function RoleOperation_CreateRoleProto.SendProto(proto)

    local ms = CS.LuaHelper.Instance:CreateMemoryStream();
    ms:WriteUShort(proto.ProtoCode);

    ms:WriteByte(proto.JobId);
    ms:WriteUTF8String(proto.RoleNickName);

    CS.LuaHelper.Instance:SendProto(ms:ToArray());
    ms:Dispose();
end


--解析协议
function RoleOperation_CreateRoleProto.GetProto(buffer)

    local proto = RoleOperation_CreateRoleProto.New(); --实例化一个协议对象
    local ms = CS.LuaHelper.Instance:CreateMemoryStream(buffer);

    proto.JobId = ms:ReadByte();
    proto.RoleNickName = ms:ReadUTF8String();

    ms:Dispose();
    return proto;
end